docs/DeveloperGuide.md

```diff
@@ -219,14 +220,39 @@ This section outlines the interactions between components of the protocol gatewa
+4. If QoS=0,
+    - `MqttIotHubAdapter` sends a PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing)
+    - At the same time `MqttIotHubAdater` completes the message on IoT hub.
+5. If QoS=1,
+    - `MqttIotHubAdapter` posts a PUBLISH packet to `RequestAckPairProcessor` responsible for handling PUBLISH-PUBACK communication.
+    - `RequestAckPairProcessor` sends PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing).
+    - at the same time `RequestAckPairProcessor` enqueues an entry with the message state (packet identifier) in its internal ACK-pending queue. The ACK-pending queue is used for tracking messages in-flight that await on acknowledgement message to arrive from the device.
+    - Eventually (assuming normal operation), PUBACK packet corresponding to the PUBLISH packet is received from the device and gets through the pipeline to `MqttIotHubAdapter`.
+    - `MqttIotHubAdapter` dispatches PUBACK packet to `RequestAckPairProcessor`.
+    - `RequestAckPairProcessor` peeks at the top message from its ACK-pending queue, and if it does not match the received PUBACK packet, PUBACK packet gets discarded and processing stops (following steps are skipped).
+    - `RequestAckPairProcessor` dequeues the message from the ACK-pending queue and calls `MqttIotHubAdapter` to complete acknowledgement.
+    - `MqttIotHubAdapter` completes the message on IoT hub.
+5. If QoS=2,
+    - `MqttIotHubAdapter` posts a PUBLISH packet to `RequestAckPairProcessor` responsible for handling PUBLISH-PUBREC communication.
+    - `RequestAckPairProcessor` sends PUBLISH packet through the pipeline (see steps 13-16 above for CONNECT processing).
+    - at the same time `RequestAckPairProcessor` enqueues an entry with the message state (packet identifier) in its internal ACK-pending queue.
+    - Eventually (assuming normal operation), PUBREC packet corresponding to the PUBLISH packet is received from the device and gets through the pipeline to `MqttIotHubAdapter`.
+    - `MqttIotHubAdapter` dispatches PUBREC packet to `RequestAckPairProcessor`.
+    - `RequestAckPairProcessor` peeks at the top message from its ACK-pending queue, and if it does not match the received PUBREC packet, PUBREC packet gets discarded and processing stops (following steps are skipped).
+    - `RequestAckPairProcessor` dequeues the message from the ACK-pending queue and calls `MqttIotHubAdapter` to complete acknowledgement.
+    - `MqttIotHubAdapter` calls `IQos2StatePersistenceProvider.SetMessageAsync` to persist the fact of acknowledged PUBLISH packet delivery.
+    - `MqttIotHubAdapter` posts a PUBREL packet to `RequestAckPairProcessor` responsible for handling PUBREL-PUBCOMP communication.
+    - `RequestAckPairProcessor` sends PUBREL packet through the pipeline (see steps 13-16 above for CONNECT processing).
+    - at the same time `RequestAckPairProcessor` enqueues an entry with the message state in its internal ACK-pending queue.
```

Comment:
Please capitalize the first word.  Also there's a missing article, so it should be "At the same time the `RequestAckPairProcessor`..."